src/sagemaker_xgboost_container/encoder.py

```diff
@@ -68,10 +72,39 @@ def libsvm_to_dmatrix(string_like):  # type: (bytes) -> xgb.DMatrix
     return dmatrix
 
 
+def recordio_protobuf_to_dmatrix(string_like):  # type: (bytes) -> xgb.DMatrix
+    """Convert a RecordIO-Protobuf byte representation to a DMatrix object.
+    Args:
+        string_like (bytes): RecordIO-Protobuf bytes.
+    Returns:
+    (xgb.DMatrix): XGBoost DataMatrix
+    """
+    buf = bytes(string_like)
+    dataset = [mlio.InMemoryStore(buf)]
+    reader = mlio.RecordIOProtobufReader(dataset=dataset, batch_size=100)
+
+    if type(reader.peek_example()['values']) is mlio.core.DenseTensor:
+        to_matrix = as_numpy
+        vstack = np.vstack
+    else:
+        to_matrix = to_coo_matrix
```

Comment:
The `to_coo_matrix` above will create a copy of the data whereas the `as_numpy` will not. 
@cbalioglu - is the above statement correct? 

@aws-patlin - There might not be anything we can change here but useful to understand the flow.

Reply:
Either way, all the data needs to be copied when the data is stacked, and then copied again when passed to DMatrix. I think the total memory usage will end up the same.

Reply:
@cbalioglu Is it possible to "append" examples in a contiguous chunk while reading and then convert using `as_numpy`?
Alternatively is there another solution in ML-IO to the below usecase of 'append and vstack' - that seems to be a common usecase and is performing an unnecessary copy.

Reply:
I think there's a real challenge here. Afaik, we don't have a way to know the total size or shape of the data before we read through it, so it's not possible for us to allocate ahead of time the amount of memory we need. Because of that, I don't think we can guarantee that the data we've read is contiguous in memory, thus requiring a copy when we stack the data together.

Reply:
@iyerr3 Correct, `to_coo_matrix` causes allocation while `as_numpy` is a zero-copy operation. Regarding your second question; I am planning to have some convenience functions that have a similar API to Pandas (e.g. `read_csv`)